Compilation fail on latest master

In future_overhead test, compilation fails because
`HPX_PLAIN_ACTION(null_function, null_action)`
causes some kind of problem.
This only happens for me on daint and might be caused by some oddity in my configuration
```
[100%] Building CXX object tests/performance/local/CMakeFiles/future_overhead_test.dir/future_overhead.cpp.o
In file included from /apps/daint/UES/biddisco/src/hpx/hpx/lcos/base_lco.hpp:11:0,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/lcos/base_lco_with_value.hpp:13,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/agas/server/primary_namespace.hpp:15,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/applier/detail/apply_colocated.hpp:15,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/include/apply.hpp:11,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:11:
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/basic_action.hpp: In instantiation of 'struct hpx::actions::make_action<double (*)() noexcept, null_function, null_action>':
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:126:1:   required from here
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/basic_action.hpp:561:12: error: invalid use of incomplete type 'struct hpx::actions::action<double (*)() noexcept, null_function, null_action>'
     struct make_action
            ^~~~~~~~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/basic_action.hpp:531:12: note: declaration of 'struct hpx::actions::action<double (*)() noexcept, null_function, null_action>'
     struct action;
            ^~~~~~
In file included from /apps/daint/UES/biddisco/src/hpx/libs/errors/include/hpx/errors/throw_exception.hpp:17:0,
                 from /apps/daint/UES/biddisco/src/hpx/libs/errors/include/hpx/errors/exception_fwd.hpp:63,
                 from /apps/daint/UES/biddisco/src/hpx/libs/errors/include/hpx/errors/error_code.hpp:15,
                 from /apps/daint/UES/biddisco/build/hpx-release/libs/errors/include/hpx/errors.hpp:13,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/hpx_finalize.hpp:13,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/hpx_init.hpp:16,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:10:
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/plain_action.hpp:223:40: error: invalid use of incomplete type 'using type = struct hpx::actions::action<double (*)() noexcept, null_function, null_action> {aka struct hpx::actions::action<double (*)() noexcept, null_function, null_action>}'
         decltype(&func), &func, name>::type {}                                \
                                        ^
/apps/daint/UES/biddisco/src/hpx/libs/preprocessor/include/hpx/preprocessor/expand.hpp:24:26: note: in definition of macro 'HPX_PP_EXPAND'
 #define HPX_PP_EXPAND(X) X
                          ^
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/plain_action.hpp:198:5: note: in expansion of macro 'HPX_PP_EXPAND'
     HPX_PP_EXPAND(HPX_PP_CAT(                                                 \
     ^~~~~~~~~~~~~
/apps/daint/UES/biddisco/src/hpx/libs/preprocessor/include/hpx/preprocessor/cat.hpp:37:28: note: in expansion of macro 'HPX_DEFINE_PLAIN_ACTION_2'
 #define HPX_PP_CAT_I(a, b) a##b
                            ^
/apps/daint/UES/biddisco/src/hpx/libs/preprocessor/include/hpx/preprocessor/cat.hpp:29:26: note: in expansion of macro 'HPX_PP_CAT_I'
 #define HPX_PP_CAT(a, b) HPX_PP_CAT_I(a, b)
                          ^~~~~~~~~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/plain_action.hpp:198:19: note: in expansion of macro 'HPX_PP_CAT'
     HPX_PP_EXPAND(HPX_PP_CAT(                                                 \
                   ^~~~~~~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/plain_action.hpp:188:5: note: in expansion of macro 'HPX_DEFINE_PLAIN_ACTION_'
     HPX_DEFINE_PLAIN_ACTION_(__VA_ARGS__)                                     \
     ^~~~~~~~~~~~~~~~~~~~~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/plain_action.hpp:365:5: note: in expansion of macro 'HPX_DEFINE_PLAIN_ACTION'
     HPX_DEFINE_PLAIN_ACTION(func, name);                                      \
     ^~~~~~~~~~~~~~~~~~~~~~~
/apps/daint/UES/biddisco/src/hpx/libs/preprocessor/include/hpx/preprocessor/cat.hpp:37:28: note: in expansion of macro 'HPX_PLAIN_ACTION_2'
 #define HPX_PP_CAT_I(a, b) a##b
                            ^
/apps/daint/UES/biddisco/src/hpx/libs/preprocessor/include/hpx/preprocessor/cat.hpp:29:26: note: in expansion of macro 'HPX_PP_CAT_I'
 #define HPX_PP_CAT(a, b) HPX_PP_CAT_I(a, b)
                          ^~~~~~~~~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/plain_action.hpp:360:19: note: in expansion of macro 'HPX_PP_CAT'
     HPX_PP_EXPAND(HPX_PP_CAT(                                                 \
                   ^~~~~~~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/plain_action.hpp:293:5: note: in expansion of macro 'HPX_PLAIN_ACTION_'
     HPX_PLAIN_ACTION_(__VA_ARGS__)                                            \
     ^~~~~~~~~~~~~~~~~
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:126:1: note: in expansion of macro 'HPX_PLAIN_ACTION'
 HPX_PLAIN_ACTION(null_function, null_action)
 ^
In file included from /apps/daint/UES/biddisco/src/hpx/hpx/lcos/base_lco.hpp:11:0,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/lcos/base_lco_with_value.hpp:13,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/agas/server/primary_namespace.hpp:15,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/applier/detail/apply_colocated.hpp:15,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/include/apply.hpp:11,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:11:
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/basic_action.hpp:531:12: note: declaration of 'using type = struct hpx::actions::action<double (*)() noexcept, null_function, null_action> {aka struct hpx::actions::action<double (*)() noexcept, null_function, null_action>}'
     struct action;
            ^~~~~~
In file included from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/action_priority.hpp:14:0,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/applier/apply.hpp:13,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/apply.hpp:11,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/include/apply.hpp:10,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:11:
/apps/daint/UES/biddisco/src/hpx/hpx/traits/extract_action.hpp: In instantiation of 'struct hpx::traits::extract_action<null_action, void>':
/apps/daint/UES/biddisco/src/hpx/hpx/lcos/async.hpp:145:9:   required from 'struct hpx::detail::async_action_dispatch<null_action, hpx::naming::id_type, void>'
/apps/daint/UES/biddisco/src/hpx/hpx/lcos/async_fwd.hpp:42:20:   required by substitution of 'template<class Action, class F, class ... Ts> decltype (hpx::detail::async_action_dispatch<Action, typename hpx::util::decay<F>::type>::call(forward<F>(f), (forward<Ts>)(hpx::async::ts)...)) hpx::async(F&&, Ts&& ...) [with Action = null_action; F = const hpx::naming::id_type&; Ts = {}]'
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:146:50:   required from here
/apps/daint/UES/biddisco/src/hpx/hpx/traits/extract_action.hpp:17:51: error: no type named 'derived_type' in 'struct null_action'
         using type = typename Action::derived_type;
                                                   ^
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp: In function 'void measure_action_futures_wait_each(uint64_t, bool)':
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:146:50: error: no matching function for call to 'async<null_action>(const hpx::naming::id_type&)'
         futures.push_back(async<null_action>(here));
                                                  ^
In file included from /apps/daint/UES/biddisco/src/hpx/hpx/include/async.hpp:10:0,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:12:
/apps/daint/UES/biddisco/src/hpx/hpx/async.hpp:107:26: note: candidate: template<class F, class ... Ts> decltype (hpx::detail::async_dispatch<typename hpx::util::decay<F>::type>::call(forward<F>(f), (forward<Ts>)(hpx::async::ts)...)) hpx::async(F&&, Ts&& ...)
     HPX_FORCEINLINE auto async(F&& f, Ts&&... ts)
                          ^~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/async.hpp:107:26: note:   template argument deduction/substitution failed:
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:146:50: note:   cannot convert 'here' (type 'const hpx::naming::id_type') to type 'null_action&&'
         futures.push_back(async<null_action>(here));
                                                  ^
In file included from /apps/daint/UES/biddisco/src/hpx/hpx/async.hpp:12:0,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/include/async.hpp:10,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:12:
/apps/daint/UES/biddisco/src/hpx/hpx/lcos/async.hpp:228:10: note: candidate: template<class Action, class F, class ... Ts> decltype (hpx::detail::async_action_dispatch<Action, typename hpx::util::decay<F>::type>::call(forward<F>(f), (forward<Ts>)(hpx::async::ts)...)) hpx::async(F&&, Ts&& ...)
     auto async(F && f, Ts &&... ts)
          ^~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/lcos/async.hpp:228:10: note:   substitution of deduced template arguments resulted in errors seen above
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp: In function 'void measure_action_futures_wait_all(uint64_t, bool)':
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:164:50: error: no matching function for call to 'async<null_action>(const hpx::naming::id_type&)'
         futures.push_back(async<null_action>(here));
                                                  ^
In file included from /apps/daint/UES/biddisco/src/hpx/hpx/include/async.hpp:10:0,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:12:
/apps/daint/UES/biddisco/src/hpx/hpx/async.hpp:107:26: note: candidate: template<class F, class ... Ts> decltype (hpx::detail::async_dispatch<typename hpx::util::decay<F>::type>::call(forward<F>(f), (forward<Ts>)(hpx::async::ts)...)) hpx::async(F&&, Ts&& ...)
     HPX_FORCEINLINE auto async(F&& f, Ts&&... ts)
                          ^~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/async.hpp:107:26: note:   template argument deduction/substitution failed:
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:164:50: note:   cannot convert 'here' (type 'const hpx::naming::id_type') to type 'null_action&&'
         futures.push_back(async<null_action>(here));
                                                  ^
In file included from /apps/daint/UES/biddisco/src/hpx/hpx/async.hpp:12:0,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/include/async.hpp:10,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:12:
/apps/daint/UES/biddisco/src/hpx/hpx/lcos/async.hpp:228:10: note: candidate: template<class Action, class F, class ... Ts> decltype (hpx::detail::async_action_dispatch<Action, typename hpx::util::decay<F>::type>::call(forward<F>(f), (forward<Ts>)(hpx::async::ts)...)) hpx::async(F&&, Ts&& ...)
     auto async(F && f, Ts &&... ts)
          ^~~~~
/apps/daint/UES/biddisco/src/hpx/hpx/lcos/async.hpp:228:10: note:   template argument deduction/substitution failed:
In file included from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/action_priority.hpp:11:0,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/applier/apply.hpp:13,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/apply.hpp:11,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/include/apply.hpp:10,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:11:
/apps/daint/UES/biddisco/src/hpx/hpx/lcos/async_fwd.hpp: In substitution of 'template<class Action, class F, class ... Ts> decltype (hpx::detail::async_action_dispatch<Action, typename hpx::util::decay<F>::type>::call(forward<F>(f), (forward<Ts>)(hpx::async::ts)...)) hpx::async(F&&, Ts&& ...) [with Action = null_action; F = const hpx::naming::id_type&; Ts = {}]':
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:164:50:   required from here
/apps/daint/UES/biddisco/src/hpx/hpx/lcos/async_fwd.hpp:42:20: error: 'call' is not a member of 'hpx::detail::async_action_dispatch<null_action, hpx::naming::id_type, void>'
     ->  decltype(detail::async_action_dispatch<
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                 Action, typename util::decay<F>::type
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             >::call(std::forward<F>(f), std::forward<Ts>(ts)...)
             ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /apps/daint/UES/biddisco/src/hpx/hpx/lcos/base_lco.hpp:11:0,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/lcos/base_lco_with_value.hpp:13,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/agas/server/primary_namespace.hpp:15,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/runtime/applier/detail/apply_colocated.hpp:15,
                 from /apps/daint/UES/biddisco/src/hpx/hpx/include/apply.hpp:11,
                 from /apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:11:
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/basic_action.hpp: In instantiation of 'void hpx::actions::detail::register_local_action_invocation_count(hpx::actions::detail::invocation_count_registry&) [with Action = null_action]':
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/detail/invocation_count_registry.hpp:79:59:   required from 'hpx::actions::detail::register_action_invocation_count<Action>::register_action_invocation_count() [with Action = null_action]'
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/detail/invocation_count_registry.hpp:93:9:   required from 'hpx::actions::detail::register_action_invocation_count<null_action> hpx::actions::detail::register_action_invocation_count<null_action>::instance'
/apps/daint/UES/biddisco/src/hpx/tests/performance/local/future_overhead.cpp:126:1:   required from here
/apps/daint/UES/biddisco/src/hpx/hpx/runtime/actions/basic_action.hpp:504:17: error: 'get_invocation_count' is not a member of 'null_action'
                 &Action::get_invocation_count
                 ^~~~~~~
make[3]: *** [tests/performance/local/CMakeFiles/future_overhead_test.dir/build.make:63: tests/performance/local/CMakeFiles/future_overhead_test.dir/future_overhead.cpp.o] Error 1
make[3]: Target 'tests/performance/local/CMakeFiles/future_overhead_test.dir/build' not remade because of errors.
make[2]: *** [CMakeFiles/Makefile2:53946: tests/performance/local/CMakeFiles/future_overhead_test.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:53953: tests/performance/local/CMakeFiles/future_overhead_test.dir/rule] Error 2
make[1]: Target 'future_overhead_test' not remade because of errors.
```

--- a/cmake/HPX_PerformCxxFeatureTests.cmake
+++ b/cmake/HPX_PerformCxxFeatureTests.cmake
@@ -228,8 +228,12 @@ function(hpx_perform_cxx_feature_tests)
     hpx_check_for_cxx17_std_in_place_type_t(
       DEFINITIONS HPX_HAVE_CXX17_STD_IN_PLACE_TYPE_T)
 
-    hpx_check_for_cxx17_noexcept_functions_as_nontype_template_arguments(
-      DEFINITIONS HPX_HAVE_CXX17_NOEXCEPT_FUNCTIONS_AS_NONTYPE_TEMPLATE_ARGUMENTS)
-
   endif()
+
+  # we deliberately check for this functionality even for non-C++17
+  # configurations as some compilers (notable gcc V7.x) require for noexcept
+  # function specializations for actions even in C++11/14 mode
+  hpx_check_for_cxx17_noexcept_functions_as_nontype_template_arguments(
+    DEFINITIONS HPX_HAVE_CXX17_NOEXCEPT_FUNCTIONS_AS_NONTYPE_TEMPLATE_ARGUMENTS)
+
 endfunction()

